internal/storage/blob/store.go

```diff
@@ -189,95 +185,133 @@ func NewStore(ctx context.Context, conf *Conf, cloner bucketCloner) (*Store, err
-		return err
+		return fmt.Errorf("failed to build index from the new set of files: %w", err)
 	}
 
+	s.workDir = newWorkDir
+	s.workFS = os.DirFS(newWorkDir)
+	s.idx = idx
 	go s.pollForUpdates(ctx)
 
 	return nil
 }
 
-func (s *Store) clone(ctx context.Context) (*CloneResult, error) {
-	ctx, cancelFunc := s.conf.getCloneCtx(ctx)
-	defer cancelFunc()
-
-	return s.cloner.Clone(ctx)
-}
-
 func (s *Store) updateIndex(ctx context.Context) error {
 	s.log.Debug("Checking for updates")
-
-	changes, err := s.clone(ctx)
+	cr, err := s.clone(ctx)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to clone blob store: %w", err)
 	}
 
-	if failures := changes.failures(); failures > 0 {
-		s.log.Warnf("Failed to download (%d) files", failures)
+	if cr.failuresCount > 0 {
+		s.log.Warnf("Failed to download (%d) files", cr.failuresCount)
+		return ErrPartialFailureToDownload
 	}
 
-	if changes.isEmpty() {
+	if cr.isEmpty() {
 		s.log.Debug("No changes")
 		return nil
 	}
 
-	s.log.Infof("Detected changes: added or updated (%d), deleted (%d)", len(changes.updateOrAdd), len(changes.delete))
+	s.log.Infof("Detected changes: added or updated (%d), deleted (%d)", len(cr.addedOrUpdated), len(cr.deleted))
 
-	var p *policyv1.Policy
-	var event storage.Event
-	for _, f := range changes.updateOrAdd {
-		if schemaFile, ok := util.RelativeSchemaPath(f.file); ok {
-			s.NotifySubscribers(storage.NewSchemaEvent(storage.EventAddOrUpdateSchema, schemaFile))
-			continue
-		}
+	idx, newWorkDir, err := s.buildIndex(ctx, cr.all)
+	if err != nil {
+		return fmt.Errorf("failed to build index from the new set of files: %w", err)
+	}
 
-		p, err = policy.ReadPolicyFromFile(s.fsys, f.file)
-		if err != nil {
-			return err
-		}
-		entry := index.Entry{File: f.file, Policy: policy.Wrap(policy.WithSourceAttributes(p, driverSourceAttr, etagSourceAttr(f.etag)))}
-		event, err = s.idx.AddOrUpdate(entry)
+	oldWorkDir := s.workDir
+	s.workDir = newWorkDir
+	s.workFS = os.DirFS(newWorkDir)
+	s.idx = idx
+
+	evts := make([]storage.Event, 0, len(cr.addedOrUpdated)+len(cr.deleted))
+	for _, i := range cr.addedOrUpdated {
+		e, err := s.addOrUpdateEvent(i.etag, i.file)
 		if err != nil {
-			return err
+			return fmt.Errorf("faield to create add or update event: %w", err)
 		}
-		s.NotifySubscribers(event)
+		evts = append(evts, e)
 	}
 
-	for _, f := range changes.delete {
-		if schemaFile, ok := util.RelativeSchemaPath(f); ok {
-			s.NotifySubscribers(storage.NewSchemaEvent(storage.EventDeleteSchema, schemaFile))
-			continue
-		}
-
-		entry := index.Entry{File: f}
-		event, err = s.idx.Delete(entry)
+	for _, i := range cr.deleted {
+		e, err := s.deleteEvent(i.etag, i.file)
 		if err != nil {
-			return err
+			return fmt.Errorf("faield to create delete event: %w", err)
 		}
-		s.NotifySubscribers(event)
+		evts = append(evts, e)
+	}
+
+	if err := s.cloner.Clean(); err != nil {
+		return fmt.Errorf("failed to clean up the cache: %w", err)
```

Comment:
We should log this as a warning but I don't think we need to fail the whole thing by returning the error because failing to clean up the files doesn't affect the correct functioning of the store.